# Unsupervised Clustering Methods for Meteorological European Configurations/ Patterns


<span style="color: yellow;"> - prendo solo un sub set dei dati per la velocità (fatto in -> 1.1), poi sarà da prendre tutto il dataset (30.07.25)  </span>  
<span style="color: yellow;">- vedere se togliere i percentili in 1.1</span>

In [1]:
import xarray as xr
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

from sklearn.preprocessing import StandardScaler, RobustScaler
from sklearn.decomposition import PCA
from sklearn.feature_selection import VarianceThreshold
from statsmodels.stats.outliers_influence import variance_inflation_factor


## 1 Caricamento Dati e Analisi Iniziale

In [2]:
try:
    ds = xr.open_dataset('era5_2000_2004.grib', engine= 'cfgrib') # XArray DataSet
    print("Dataset loaded successfully.")
except Exception as e:
    print(f"Error loading dataset: {e}")

Dataset loaded successfully.


In [3]:
print("Overview of the dataset:")
print(f"   • Variabili: {list(ds.data_vars.keys())}")
print(f"   • Coordinate: {list(ds.coords.keys())}")

Overview of the dataset:
   • Variabili: ['z', 't', 'u', 'v']
   • Coordinate: ['number', 'time', 'step', 'isobaricInhPa', 'latitude', 'longitude', 'valid_time']


In [4]:
# Dimenision details
print("Dimension details:")
if 'latitude' in ds.dims:
    print(f"   • Latitude: {ds.dims['latitude']} points ({ds.latitude.min().values:.1f}° - {ds.latitude.max().values:.1f}°)")
if 'longitude' in ds.dims:
    print(f"   • Longitude: {ds.dims['longitude']} points ({ds.longitude.min().values:.1f}° - {ds.longitude.max().values:.1f}°)")
if 'time' in ds.dims:
    print(f"   • Time: {ds.dims['time']} steps ({pd.to_datetime(ds.time.values[0]).strftime('%Y-%m-%d')} - {pd.to_datetime(ds.time.values[-1]).strftime('%Y-%m-%d')})")
if 'isobaricInhPa' in ds.dims:
    print(f"   • Pressure levels: {ds.dims['isobaricInhPa']} levels ({list(ds.isobaricInhPa.values)} hPa)")

#Variables 
print("Variables in the dataset:")
for var in ds.data_vars:
    var_data = ds[var]
    print(f"   • {var}: {var_data.dims} - {var_data.attrs.get('long_name', 'N/A')}")
    print(f"     └─ Units: {var_data.attrs.get('units', 'N/A')}")


Dimension details:
   • Latitude: 201 points (20.0° - 70.0°)
   • Longitude: 321 points (-40.0° - 40.0°)
   • Time: 1827 steps (2000-01-01 - 2004-12-31)
   • Pressure levels: 3 levels ([np.float64(850.0), np.float64(500.0), np.float64(250.0)] hPa)
Variables in the dataset:
   • z: ('time', 'isobaricInhPa', 'latitude', 'longitude') - Geopotential
     └─ Units: m**2 s**-2
   • t: ('time', 'isobaricInhPa', 'latitude', 'longitude') - Temperature
     └─ Units: K
   • u: ('time', 'isobaricInhPa', 'latitude', 'longitude') - U component of wind
     └─ Units: m s**-1
   • v: ('time', 'isobaricInhPa', 'latitude', 'longitude') - V component of wind
     └─ Units: m s**-1


In [5]:
# Total dimensionality
total_spatial_points = 1
for dim in ['latitude', 'longitude']:
    if dim in ds.dims:
        total_spatial_points *= ds.dims[dim]

total_features = len(ds.data_vars) * ds.dims.get('isobaricInhPa', 1) * total_spatial_points
print("DIMENSIONALITY:")
print(f"   • Spatial points: {total_spatial_points}")
print(f"   • Total features per timestep: {total_features:,}")
print(f"   • Temporal samples: {ds.dims.get('time', 1)}")

DIMENSIONALITY:
   • Spatial points: 64521
   • Total features per timestep: 774,252
   • Temporal samples: 1827


Spatial points: punti griglia nello spazio. 
La regione osservata è suddivisa in una griglia regolare (0.25° x 0.25°), per ogni punto nella grigli avengono misurate le variabili

Total features per timestep: numero di variabili (features) in totale in ogni istante di tempo

Temporal samples: punti temporali nel dataset (365 giorni per 5 anni)

Posso trasformarlo in una matrice per il clustering:  
shape = (temporal_samples, total_features_per_timestep)
       = (1827, 774252)  
Ogni riga = una mappa meteorologica


### 1.1 Check quality in data

In [6]:
# Select a subset of the dataset for a specific time range (POI DA TOGLIERE)
#ds = ds.sel(time=slice("2000-01-01", "2001-12-31"))

In [7]:
def analyze_missing_values(dataset):
    """Analizzes missing values for each variable"""
    missing_info = {}
    
    for var in dataset.data_vars:
        data = dataset[var]
        total_values = data.size
        missing_count = np.isnan(data.values).sum()
        missing_percent = (missing_count / total_values) * 100
        
        missing_info[var] = {
            'count': missing_count,
            'percentage': missing_percent,
            'total': total_values
        }
    
    return missing_info

print("MISSING VALUES:")
missing_analysis = analyze_missing_values(ds)

for var, info in missing_analysis.items():
    print(f"    {var}: {info['count']:,} missing ({info['percentage']:.2f}%)")

MISSING VALUES:
    z: 0 missing (0.00%)
    t: 0 missing (0.00%)
    u: 0 missing (0.00%)
    v: 0 missing (0.00%)


There are no missing values in the dataset

In [8]:
print("STATISTICS:")
for var in ds.data_vars:
    data = ds[var].values
    valid_data = data[~np.isnan(data)]
    
    if len(valid_data) > 0:
        print(f" {var.upper()}:")
        print(f"      • Min: {valid_data.min():.3f}")
        print(f"      • Max: {valid_data.max():.3f}")
        print(f"      • Mean: {valid_data.mean():.3f}")
        print(f"      • Std: {valid_data.std():.3f}")
        print(f"      • Percentiles [25%, 50%, 75%]: {np.percentile(valid_data, [25, 50, 75])}")

STATISTICS:
 Z:
      • Min: 7495.887
      • Max: 108808.312
      • Mean: 57405.691
      • Std: 36090.105
      • Percentiles [25%, 50%, 75%]: [ 15097.1484375  55659.46875   100139.4375   ]
 T:
      • Min: 198.893
      • Max: 308.547
      • Mean: 252.359
      • Std: 24.976
      • Percentiles [25%, 50%, 75%]: [226.2230072  255.91355896 272.93847656]
 U:
      • Min: -63.877
      • Max: 112.808
      • Mean: 8.984
      • Std: 13.904
      • Percentiles [25%, 50%, 75%]: [-0.31930542  6.58242798 16.01531982]
 V:
      • Min: -91.157
      • Max: 89.342
      • Mean: -0.238
      • Std: 11.834
      • Percentiles [25%, 50%, 75%]: [-5.98008728 -0.32400513  5.86112976]


## 2 Preprocessing and Feature Selection

Ho 4 variabili: z, u, v, t
Per ogni variabile: var[time, pressure, latitude, longitude]

Ogni variabile ha 64521 valori per timestamp (3 × 201 × 321 = 64521 punti spaziali)(3 lv. di pressione x 201 lat x 321 lon)  
Con 4 variabili ⇒ ogni riga della matrice finale avrà:

    4 × 64521 = 774252 colonne (features)

Un timestamp è come un pacco 3D: (3, 201, 321, 4)
Un pacco per ogni giorno dal 01/01/2000 al 31/12/2004 → 1827 pacchi

#### Struttura inziale dei dati

**Scatola** = Dataset   
dentro la scatola di sono dei blocchi di fogli

**Un blocco di fogli** = un signolo giorno ( da 1 gennaio 2000 a 21 dic 2004)  -> 1827 giorni  
il blocco di fogli è formato da 4 fogli uno per ogni variabile

**Un foglio contiene i valori di una variabile** =  variabili: u, v, z, t -> 4 variabili  

Ogni foglio contiene i valori di quella variabile presi in ogni singolo punto dello 'spazio' definito dalla longitudine e dalla laitudine. Quindi in ogni foglio c'è il valore di quella variabile in ognuno dei 201(lat) × 321(lon). Una specie di tabella.  -> 64521 punti spaziali

**Solo che questa tabella di valori è presa per ognuno dei 3 livelli di pressione** = 850 hPa, 500 hPa, 250 hPa -> 3 lv di pressione

**TOT= 774252 valori per blocco**   x 1827 giorni

Per ogni variabile:  
__per ogni livello di pressione:  
____per ogni lat:  
______per ogni lon:  
________prendi il valore  

Immagina il foglio come una tabella con 774252 colonne, e solo 1 riga, che rappresenta tutte le misure spaziali per quel giorno.
Se metti insieme tutti i 1827 fogli, ottieni una matrice finale di forma (1827, 774252). Ogni riga è un giorno. Ogni colonna è una variabile a una certa posizione e pressione.


_capire come vengono modificate le dimensioni nella modifica fatta, 2 chat su chat gpt_


L’obiettivo è trasformare tutto in una tabella 2D:

           feature_1  feature_2  ...  feature_774252
time_1 →      ...        ...             ...
time_2 →      ...        ...             ...
  ⋮                             
time_1827 →   ...        ...             ...

    Righe: 1827 (una per ogni timestep)

    Colonne: 774,252 (una per ogni combinazione di punto spaziale × variabile)

    Ogni colonna è, ad esempio:
    "z_850hPa_lat37.5_lon12.0"
    "u_500hPa_lat42.5_lon18.0"

In [9]:
# Converti dataset in array 2D per ML
def prepare_data_matrix(dataset):
    """Converte xarray dataset in matrice 2D per ML"""
    data_matrices = {}
    
    for var in dataset.data_vars:
        print(f"   • Processando {var}...")
        var_data = dataset[var]
        
        # Riorganizza dimensioni: (time, features)
        if 'time' in var_data.dims:
            # Stack tutte le dimensioni non temporali
            spatial_dims = [dim for dim in var_data.dims if dim != 'time']
            if spatial_dims:
                stacked = var_data.stack(features=spatial_dims)
                matrix = stacked.values  # shape: (time, features)
            else:
                matrix = var_data.values.reshape(-1, 1)
        else:
            matrix = var_data.values.flatten().reshape(1, -1)
        
        data_matrices[var] = matrix
    
    # Concatena tutte le variabili
    all_matrices = list(data_matrices.values())
    combined_matrix = np.concatenate(all_matrices, axis=1)
    
    # Rimuovi NaN
    mask = ~np.isnan(combined_matrix).any(axis=1)
    clean_matrix = combined_matrix[mask]
    
    return clean_matrix, data_matrices

X, data_matrices = prepare_data_matrix(ds)
print(f"✅ Matrice dati: {X.shape} (samples, features)")

   • Processando z...
   • Processando t...
   • Processando u...
   • Processando v...


KeyboardInterrupt: 

In [10]:
201*321

64521